File: app/src/main/java/fr/free/nrw/commons/wikidata/mwapi/MwQueryPage.java
package fr.free.nrw.commons.wikidata.mwapi;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.gson.annotations.SerializedName;

import org.apache.commons.lang3.StringUtils;
import fr.free.nrw.commons.wikidata.model.gallery.ImageInfo;
import fr.free.nrw.commons.wikidata.model.BaseModel;

import java.util.Collections;
import java.util.List;

/**
 * A class representing a standard page object as returned by the MediaWiki API.
 */
public class MwQueryPage extends BaseModel {
    private int pageid;
    private int index;
    @NonNull private String title;
    @NonNull private CategoryInfo categoryinfo;
    @Nullable private List<Revision> revisions;
    @SerializedName("fileusage") @Nullable private List<FileUsage> fileUsages;
    @SerializedName("globalusage") @Nullable private List<GlobalUsage> globalUsages;
    @Nullable private List<Coordinates> coordinates;
    @Nullable private List<Category> categories;
    @Nullable private Thumbnail thumbnail;
    @Nullable private String description;
    @SerializedName("imageinfo") @Nullable private List<ImageInfo> imageInfo;
    @Nullable private String redirectFrom;
    @Nullable private String convertedFrom;
    @Nullable private String convertedTo;

    @NonNull public String title() {
        return title;
    }

    @NonNull public CategoryInfo categoryInfo() {
        return categoryinfo;
    }

    public int index() {
        return index;
    }

    @Nullable public List<Revision> revisions() {
        return revisions;
    }

    @Nullable public List<Category> categories() {
        return categories;
    }

    @Nullable public List<Coordinates> coordinates() {
        // TODO: Handle null values in lists during deserialization, perhaps with a new
        // @RequiredElements annotation and corresponding TypeAdapter
        if (coordinates != null) {
            coordinates.removeAll(Collections.singleton(null));
        }
        return coordinates;
    }

    public int pageId() {
        return pageid;
    }

    @Nullable public String thumbUrl() {
        return thumbnail != null ? thumbnail.source() : null;
    }

    @Nullable public String description() {
        return description;
    }

    @Nullable public ImageInfo imageInfo() {
        return imageInfo != null ? imageInfo.get(0) : null;
    }

    public void redirectFrom(@Nullable String from) {
        redirectFrom = from;
    }

    public void convertedFrom(@Nullable String from) {
        convertedFrom = from;
    }

    public void convertedTo(@Nullable String to) {
        convertedTo = to;
    }

    public void appendTitleFragment(@Nullable String fragment) {
        title += "#" + fragment;
    }

    public boolean checkWhetherFileIsUsedInWikis() {
        if (globalUsages != null && globalUsages.size() > 0) {
            return true;
        }

        if (fileUsages == null || fileUsages.size() == 0) {
            return false;
        }

        final int totalCount = fileUsages.size();

        /* Ignore usage under https://commons.wikimedia.org/wiki/User:Didym/Mobile_upload/
           which has been a gallery of all of our uploads since 2014 */
        for (final FileUsage fileUsage : fileUsages) {
            if ( ! fileUsage.title().contains("User:Didym/Mobile upload")) {
                return true;
            }
        }

        return false;
    }

    public static class Revision {
        @SerializedName("revid") private long revisionId;
        private String user;
        @SerializedName("contentformat") @NonNull private String contentFormat;
        @SerializedName("contentmodel") @NonNull private String contentModel;
        @SerializedName("timestamp") @NonNull private String timeStamp;
        @NonNull private String content;

        @NonNull public String content() {
            return content;
        }

        @NonNull public String timeStamp() {
            return StringUtils.defaultString(timeStamp);
        }

        public long getRevisionId() {
            return revisionId;
        }

        @NonNull
        public String getUser() {
            return StringUtils.defaultString(user);
        }
    }

    public static class Coordinates {
         @Nullable private Double lat;
         @Nullable private Double lon;

        @Nullable public Double lat() {
            return lat;
        }
        @Nullable public Double lon() {
            return lon;
        }
    }

    public static class CategoryInfo {
         private boolean hidden;
         private int size;
         private int pages;
         private int files;
         private int subcats;
        public boolean isHidden() {
            return hidden;
        }
    }

    static class Thumbnail {
         private String source;
         private int width;
         private int height;
        String source() {
            return source;
        }
    }

    public static class GlobalUsage {
        @SerializedName("title") private String title;
        @SerializedName("wiki")private String wiki;
        @SerializedName("url") private String url;

        public String getTitle() {
            return title;
        }

        public String getWiki() {
            return wiki;
        }

        public String getUrl() {
            return url;
        }
    }

    public static class FileUsage {
        @SerializedName("pageid") private int pageid;
        @SerializedName("ns") private int ns;
        @SerializedName("title") private String title;

        public int pageId() {
            return pageid;
        }

        public int ns() {
            return ns;
        }

        public String title() {
            return title;
        }
    }

    public static class Category {
         private int ns;
        @SuppressWarnings("unused,NullableProblems") @Nullable private String title;
         private boolean hidden;

        public int ns() {
            return ns;
        }

        @NonNull public String title() {
            return StringUtils.defaultString(title);
        }

        public boolean hidden() {
            return hidden;
        }
    }
}

File: app/src/main/java/fr/free/nrw/commons/wikidata/mwapi/MwQueryPage.kt
package fr.free.nrw.commons.wikidata.mwapi

import androidx.annotation.VisibleForTesting
import com.google.gson.annotations.SerializedName
import fr.free.nrw.commons.wikidata.model.BaseModel
import fr.free.nrw.commons.wikidata.model.gallery.ImageInfo
import fr.free.nrw.commons.wikidata.mwapi.MwQueryPage.GlobalUsage

/**
 * A class representing a standard page object as returned by the MediaWiki API.
 */
class MwQueryPage : BaseModel() {
    private val pageid = 0
    private val index = 0
    private var title: String? = null
    private val categoryinfo: CategoryInfo? = null
    private val revisions: List<Revision>? = null

    @SerializedName("fileusage")
    private val fileUsages: List<FileUsage>? = null

    @SerializedName("globalusage")
    private val globalUsages: List<GlobalUsage>? = null
    private val coordinates: MutableList<Coordinates?>? = null
    private val categories: List<Category>? = null
    private val thumbnail: Thumbnail? = null
    private val description: String? = null

    @SerializedName("imageinfo")
    private val imageInfo: List<ImageInfo>? = null
    private var redirectFrom: String? = null
    private var convertedFrom: String? = null
    private var convertedTo: String? = null

    fun title(): String = title!!

    fun categoryInfo(): CategoryInfo? = categoryinfo

    fun index(): Int = index

    fun revisions(): List<Revision>? = revisions

    fun categories(): List<Category>? = categories

    // TODO: Handle null values in lists during deserialization, perhaps with a new
    // @RequiredElements annotation and corresponding TypeAdapter
    fun coordinates(): List<Coordinates?>? = coordinates?.filterNotNull()

    fun pageId(): Int = pageid

    fun thumbUrl(): String? = thumbnail?.source()

    fun description(): String? = description

    fun imageInfo(): ImageInfo? = imageInfo?.get(0)

    fun redirectFrom(from: String?) {
        redirectFrom = from
    }

    fun convertedFrom(from: String?) {
        convertedFrom = from
    }

    fun convertedTo(to: String?) {
        convertedTo = to
    }

    fun appendTitleFragment(fragment: String?) {
        title += "#$fragment"
    }

    fun checkWhetherFileIsUsedInWikis(): Boolean {
        return checkWhetherFileIsUsedInWikis(globalUsages, fileUsages)
    }

    class Revision {
        @SerializedName("revid")
        private val revisionId: Long = 0
        private val user: String? = null
        @SerializedName("contentformat")
        private val contentFormat: String? = null
        @SerializedName("contentmodel")
        private val contentModel: String? = null
        @SerializedName("timestamp")
        private val timeStamp: String? = null
        private val content: String? = null

        fun revisionId(): Long = revisionId

        fun user(): String = user ?: ""

        fun content(): String = content!!

        fun timeStamp(): String = timeStamp ?: ""
    }

    class Coordinates {
        private val lat: Double? = null
        private val lon: Double? = null

        fun lat(): Double? = lat

        fun lon(): Double? = lon
    }

    class CategoryInfo {
        val isHidden: Boolean = false
        private val size = 0
        private val pages = 0
        private val files = 0
        private val subcats = 0
    }

    internal class Thumbnail {
        private val source: String? = null
        private val width = 0
        private val height = 0

        fun source(): String? = source
    }

    class GlobalUsage {
        @SerializedName("title")
        val title: String? = null

        @SerializedName("wiki")
        val wiki: String? = null

        @SerializedName("url")
        val url: String? = null
    }

    class FileUsage {
        @SerializedName("pageid")
        private val pageid = 0

        @SerializedName("ns")
        private val ns = 0

        @SerializedName("title")
        private var title: String? = null

        fun pageId(): Int = pageid

        fun ns(): Int = ns

        fun title(): String = title ?: ""

        fun setTitle(value: String) {
            title = value
        }
    }

    class Category {
        private val ns = 0
        private val title: String? = null
        private val hidden = false

        fun ns(): Int = ns

        fun title(): String = title ?: ""

        fun hidden(): Boolean = hidden
    }
}

@VisibleForTesting
fun checkWhetherFileIsUsedInWikis(
    globalUsages: List<GlobalUsage>?,
    fileUsages: List<MwQueryPage.FileUsage>?
): Boolean {
    if (!globalUsages.isNullOrEmpty()) {
        return true
    }

    if (fileUsages.isNullOrEmpty()) {
        return false
    }

    /* Ignore usage under https://commons.wikimedia.org/wiki/User:Didym/Mobile_upload/
       which has been a gallery of all of our uploads since 2014 */
    return fileUsages.filterNot {
        it.title().contains("User:Didym/Mobile upload")
    }.isNotEmpty()
}
